File: implementations/vb/Loom.Engine/Context/MemoryManager.vb
Imports Loom.Core.Enums
Imports Loom.Core.Models

Namespace Context
    ''' <summary>
    ''' Transforms the message into a summary paragraph, extracts key data, and updates the MemoryContext based on the MemoryMode
    ''' </summary>
    Public Class MemoryManager

        Private ReadOnly _invocation As LlmInvocation

        Public Sub New(invocation As LlmInvocation)
            _invocation = invocation
        End Sub

        ''' <summary>
        ''' Synchronize memory contents according to the defined strategy
        ''' </summary>
        ''' <param name="newData"> Optional: new facts extracted externally </param>
        Public Sub UpdateMemory(Optional newData As String = "")

            Dim flag As Integer = _invocation.Memory.Mode
            ' Dovremo anche gestire la parte del ConversationManager, per ora salva lo storico della chat...
            If flag = MemoryMode.None Then
                _invocation.Memory.Content = String.Empty
            ElseIf flag = MemoryMode.FullHistory Then
                _invocation.Memory.Content &= Environment.NewLine & newData
            ElseIf flag = MemoryMode.Summary Then
                _invocation.Memory.Content = Environment.NewLine & newData
            ElseIf flag = MemoryMode.Extractive Then
                Throw New NotImplementedException("Extractive memory to implement")
            ElseIf flag = MemoryMode.Hybrid Then
                Throw New NotImplementedException("Hybrid memory to implement")
            End If

        End Sub

        ''' <summary>
        ''' Check if the memory is expired
        ''' </summary>
        ''' <param name="timestampLastUpdate"> Last update </param>
        ''' <returns></returns>
        Public Function IsMemoryExpired(timestampLastUpdate As DateTime) As Boolean

            If Not _invocation.Memory.TTL.HasValue Then Return False

            Return (DateTime.Now - timestampLastUpdate) > _invocation.Memory.TTL.Value
        End Function

    End Class
End Namespace

File: implementations/csharp/src/Loom.Engine/Context/MemoryManager.cs
using Loom.Core.Enums;
using Loom.Core.Models;
using System;


namespace Loom.Engine.Context
{

    /// <summary>
    /// Transforms the message into a summary paragraph, extracts key data, and updates the MemoryContext based on the MemoryMode
    /// </summary>
    public class MemoryManager
    {

        private readonly LlmInvocation _invocation;

        public MemoryManager(LlmInvocation invocation)
        {
            _invocation = invocation;
        }

        /// <summary>
        /// Synchronize memory contents according to the defined strategy
        /// </summary>
        /// <param name="newData">Optional: New facts extracted externally</param>
        public void UpdateMemory(string newData = "")
        {
            MemoryMode flag = _invocation.Memory.Mode;

            switch (flag)
            {
                case MemoryMode.None:
                    _invocation.Memory.Content = "";
                    break;

                case MemoryMode.FullHistory:
                    _invocation.Memory.Content += Environment.NewLine + newData;
                    break;

                case MemoryMode.Summary:
                    _invocation.Memory.Content = Environment.NewLine + newData;
                    break;

                case MemoryMode.Extractive:
                    throw new NotImplementedException("Extractive memory to implements");

                case MemoryMode.Hybrid:
                    throw new NotImplementedException("Hybrid memory to implements");
            }
        }

        /// <summary>
        ///
        /// </summary>
        /// <param name="timestampLastUpdate"></param>
        /// <returns></returns>
        public bool IsMemoryExpired(DateTime timestampLastUpdate)
        {

            if (!_invocation.Memory.TTL.HasValue) return false;

            return (DateTime.Now - timestampLastUpdate) > _invocation.Memory.TTL.Value;

        }



    }
}
